SI-6415 avoid early evaluation in some Stream's methods

jedesah opened the pull request:
The lengthCompare method in LinearSeqOptimized was looking one step further
than it needed too in order to give the correct result, which was creating some
unwanted side effects related to Streams.

https://issues.scala-lang.org/browse/SI-6415

Review by @axel22

paulp committed:
Fix for SI-6357, cycle with value classes.

Don't force the owner info.

retronym committed:
SI-6677 Insert required cast in `new qual.foo.T`

Short version: we sometimes need to rewrite this as

    new T(qual.asInstanceOf[OwnerOfFoo].foo)

Long version:

`adaptMember` in Erasure performs a few tasks, among them:

1. adding an empty argument list to qualifiers in `new qual.T`
    for which `qual` is a val template member that has
    (post uncurry) a MethodType with an empty parameter list.

    The same rewriting was already applied in uncurry for such
    qualifiers appearing in other contexts, e.g. `qual.foo` was
    already rewritten to `qual().foo`.
2. casting, if necessary, the qualifier in `Select(qual, name)`
    to the type of owner of the symbol that this selection
    references.

    This can be neccesary with compound types:
     - some val class member has type `A with B`;
     - we instantiate `new ab.valMemberOfB.T`
     - we must pass `ab.valMemberOfB` to the constructor of `T`
       as the `$outer` pointer
     - we must cast `ab` to `B` before calling `valMemberOfB`.

    Failure to insert this cast can lead to a LinkageError or
    a VerifyError.

However, if 1) was performed, 2) was not.

The error is pretty easy to trigger with the new reflection
API:

    class Test {
      val cm: reflect.runtime.universe.Mirror
          = reflect.runtime.currentMirror
      def error {
        new cm.universe.Traverser // java.lang.VerifyError
      }

      def okay1 {
        val cm: reflect.runtime.universe.Mirror = reflect.runtime.currentMirror
        new cm.universe.Traverser
      }
    }

The fix applied here to `adaptMember` mirrors the existing
implementation of `adaptType`.

retronym committed:
Refactoring of adaptMethod

Avoid the early return, and address micro-duplication
between adaptType and adaptMember.

paulp committed:
Fix for SI-6687, wrong isVar logic.

Fields which back lazy vals need to be excluded via !isLazy
lest isVar return true.

jedesah committed:
findEntry implementation code more concise and DRYer.

jedesah committed:
SI-6415 avoid early evaluation in some Stream's methods

The lengthCompare method in LinearSeqOptimized was looking one step further
than it needed to in order to give the correct result, which was creating some
unwanted side effects related to Streams.

Add test case to expose the problem and a few test cases to ensure correctness of LinearSeq.lengthCompare() method.

https://issues.scala-lang.org/browse/SI-6415

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1704/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1704/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/993/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/993/

retronym commented:
Could you please reopen this PR against the 2.10.x branch? Otherwise the fix will be delated until 2.11.0. Thanks.

axel22 commented on src/library/scala/collection/LinearSeqOptimized.scala:
```diff
@@ -246,13 +246,10 @@ trait LinearSeqOptimized[+A, +Repr <: LinearSeqOptimized[A, Repr]] extends Linea
 
   override /*SeqLike*/
```
This should be annotated with `@tailrec` now.

axel22 commented:
We need a unit test for this fix.
It should test `LinearSeq.lengthCompare` on empty streams, non-empty streams, with `len` being less, equal or greater than the stream length.
It should additionally test that the `Stream.isDefinedAt` now behaves correctly, i.e. it does not evaluate the head of the tail when called (this is what the ticket was about).

After this changes, will look good to me.

paulp committed:
Fix for SI-6706, Symbol breakage under GC.

Ensure the map key and the String in the Symbol are the
same reference by removing the old key before updating the
map with the new key -> symbol relation.

paulp committed:
Mark pattern matcher synthetics as SYNTHETIC.

Flagging synthetics accurately is important to allow for
useful static analysis.

paulp committed:
Set symbol flags at creation.

All should prefer passing flags at symbol creation to mutating
the flags field after creation.

xeno-by committed:
Merge pull request #1662 from paulp/issue/6706

Fix for SI-6706, Symbol breakage under GC.

adriaanm committed:
Merge pull request #1643 from retronym/ticket/6677

SI-6677 Insert required cast in `new qual.foo.T`

adriaanm committed:
Merge pull request #1647 from adriaanm/backport-1626-2.10.x

Fix for SI-6357, cycle with value classes.

adriaanm committed:
Merge pull request #1652 from jedesah/FlatHashTable

findEntry implementation code more concise and DRYer.

adriaanm committed:
Merge pull request #1663 from paulp/merge-2.10.wip-x

Merge 2.10.0-wip into 2.10.x.

adriaanm committed:
Merge pull request #1665 from paulp/issue/6687

Fix for SI-6687, wrong isVar logic.

adriaanm committed:
Merge pull request #1666 from paulp/issue/6707

Flag pattern matcher synthetics as synthetic.
